### PR TITLE
Use drawImage for picture layer cache

### PR DESCRIPTION
**Pull request by liyuqian**
The previous implementation uses drawImageRect with kLow_FilterQuality which will trigger bilerp inside Skia. That seems unnecessarily costly because the cache's resolution should be approximately equal to the device's resolution.

Also, the different src rect and dst rect in the previous implementation makes it a little harder to reason about the raster's behavior. (See, e.g., https://github.com/flutter/flutter/issues/17731)

This PR drops src/dst rects and simply uses translation/scale that intuitively matches (and compensates) the translation/scale of creating the cache. It also allows us to directly use drawImage so we no longer needs bilerp.

I haven't found a test that can tell the performance difference. Theoretically it can only go faster.

### CODE REVIEW

**Comment by liyuqian** [comment]
@xster FYI, I'm trying to change the cache behavior a little bit.

**Comment by xster** [comment]
This is awesome! wrt tests, in https://github.com/flutter/flutter/blob/master/dev/devicelab/README.md#running-tests-locally, I assume running `flutter_gallery__transition_perf` should show some differences locally since we rasterize leaf picture layers heuristically now anyway.

File: flow/raster_cache.cc [hunk 1]
```diff
@@ -77,6 +77,11 @@ RasterCacheResult RasterizePicture(SkPicture* picture,
 
   const SkVector3& scale = matrix.scale();
 
+  if (scale.x() == 0 || scale.y() == 0) {
```
Comment (chinmaygarde) [inline flow/raster_cache.cc]: This check is redundant as the decomposition would have failed already.

File: flow/raster_cache.cc [hunk 2]
```diff
@@ -118,8 +123,10 @@ RasterCacheResult RasterizePicture(SkPicture* picture,
 
   return {
       surface->makeImageSnapshot(),  // image
-      physical_rect,                 // source rect
-      logical_rect                   // destination rect
+      // compensate canvas->translate(-logical_rect.left(), -logical_rect.top())
```
Comment (chinmaygarde) [inline flow/raster_cache.cc]: Did you forget to run `clang-format` over your patch?

Comment (liyuqian) [reply flow/raster_cache.cc]: Yes. Running it now.

File: flow/raster_cache.cc (hunk 1 above)
Comment (liyuqian) [reply flow/raster_cache.cc]: Ok, I'll replace it with DCHECK and comment that the decomposition has already taken care of that.

**Comment by liyuqian** [comment]
@Hixie : yes, they should exactly match. This PR tries to achieve that. Our mission would be achieved  if @tvolkert 's Google3 diff returns no pixel difference.

BTW, I found a bug in my code. A new patch will be uploaded soon.

**Comment by Hixie** [comment]
I guess what I don't understand is why we have to track a scale.

**Comment by tvolkert** [comment]
@Hixie we were implicitly tracking a scale previously by virtue of maintaining a `source_rect` and `destination_rect`.  Is your point just that there shouldn't be a need to _record_ the scale since we can just extract the pixel ratio on demand when we apply the transformation from the cache?

**Comment by liyuqian** [comment]
Here's an example:

Suppose that we are drawing a 100x100 picture in a canvas with 2x scale. The cache size needed and generated is 200x200 (because of the 2x scale). That's also the screen resolution of the picture drawn. When we draw that 200x200 cache back into the the canvas, we need to first apply a 1/2 scale to the canvas to mitigate the 2x scale. Without such 1/2 scale, we would have drawn a 200x200 cache into a 400x400 region.

**Comment by liyuqian** [comment]
One more note: SkCanvas has no public API to ignore the CTM (e.g., 2x scale) and directly draw something into the device space. The underlying SkDevice has such drawSprite method. If we were using drawSprite, then we no longer need to maintain the scale (and translation). Since we're using drawImage, we need to maintain them.

**Comment by Hixie** [comment]
What happens if the transform is a shear plus a rotation?

**Comment by liyuqian** [comment]
The transformation is first decomposed into scale matrix and other matrix (e.g., shear, rotation). Only scale matrix is used to create the raster cache. The other transformation remains on the canvas' CTM, and will be applied to the raster cache when it's drawn back to the canvas.

**Comment by liyuqian** [comment]
According to @chinmaygarde , we could draw a non-rotated picture with a rotated matrix. But it seems that he'll invalidate the cache if there's perspective.

BTW, thanks to @Hixie 's comment, I realized that I've trapped myself in fixing the smaller issue of scale + offset. It turned out that if we're going to fix the whole transformation issue (e.g., rotation, perspective, etc.) altogether, we could have a simpler solution (less code, less future cognitive cost, but probably more initial cognitive cost for the current transition).

I'll prepare a new commit soon.

**Comment by liyuqian** [comment]
We have to respect the fractional part of the translation in order to draw identically between using cache and not using cache. Now I think the offset issue of the Google3 diff should be completely fixed.

Of course, the cache would be much less useful during scrolling if we have fractional translations (and we respect that for caching). Hence I rounded all translations by default (like Android). If we have some clients that have higher quality requirements and greater power (e.g., low DPI desktops), we can define SUPPORT_FRACTIONAL_TRANSLATION.

**Comment by abarth** [comment]
> the cache would be much less useful during scrolling if we have fractional translations

The scrolling code is supposed to snap scroll offsets to integer device pixels to avoid these issues.  Otherwise, you get "shimmering" effects for scrolling images as they get resampled at different fractional device pixel offsets.  (It's probably worth snapping here too given that we have actual integers here instead of the doubles we have in the scrolling code.)

**Comment by abarth** [comment]
>  It turned out that if we're going to fix the whole transformation issue (e.g., rotation, perspective, etc.) altogether, we could have a simpler solution

Yeah, this approach is probably the right choice.  I did it the other way originally because I wanted to be able to get cache hits during rotation animations, but that does cost quality.  In browsers, hitting the cache for rotations and other transforms is very important.  They even hit the cache within some bounds for scales.  However, in our setting getting device-aligned pixels is probably a better trade-off.

LGTM!

**Comment by liyuqian** [comment]
Thanks Adam!

@chinmaygarde : please let me know when you've reviewed the change. I'll merge after your LGTM :)

File: flow/raster_cache.cc [hunk 3]
```diff
@@ -70,23 +79,16 @@ static bool IsPictureWorthRasterizing(SkPicture* picture,
 
 RasterCacheResult RasterizePicture(SkPicture* picture,
                                    GrContext* context,
-                                   const MatrixDecomposition& matrix,
+                                   const SkMatrix& ctm,
```
Comment (tvolkert) [inline flow/raster_cache.cc]: Do we still need `MatrixDecomposition`?

Comment (tvolkert) [reply flow/raster_cache.cc]: I mean: can we remove the class now?

Comment (chinmaygarde) [reply flow/raster_cache.cc]: Scenic node hierarchies are updated based on the results of the matrix decomposition. That part is still in flow currently though I am planning to move it out.

Comment (liyuqian) [reply flow/raster_cache.cc]: I think that we no longer needs MatrixDecomposition. My original plan was to remove that class later so we can keep this PR small. Should I remove it in this PR?

Comment (liyuqian) [reply flow/raster_cache.cc]: Ah, just saw Chinmay's comment. Then we can only remove it after Scenic is no longer using it?

Comment (chinmaygarde) [reply flow/raster_cache.cc]: You could move it into an OS_FUCHSIA block but the unit tests for that class are only run by the engine bots. So I would keep it as is for now.

Comment (liyuqian) [reply flow/raster_cache.cc]: I see. I'll then also merge this PR as is.

**Pull request merged**